packages/block-editor/src/components/spacing-sizes-control/hooks/use-spacing-sizes.js

```diff
@@ -1,28 +1,57 @@
 /**
  * WordPress dependencies
  */
+import { useMemo } from '@wordpress/element';
 import { __ } from '@wordpress/i18n';
 
 /**
  * Internal dependencies
  */
 import { useSettings } from '../../use-settings';
+import { RANGE_CONTROL_MAX_SIZE } from '../utils';
 
 export default function useSpacingSizes() {
-	const spacingSizes = [ { name: 0, slug: '0', size: 0 } ];
+	const [
+		customSizes,
+		themeSizes,
+		defaultSizes,
+		defaultSpacingSizesEnabled,
+	] = useSettings(
+		'spacing.spacingSizes.custom',
+		'spacing.spacingSizes.theme',
+		'spacing.spacingSizes.default',
+		'spacing.defaultSpacingSizes'
+	);
 
-	const [ settingsSizes ] = useSettings( 'spacing.spacingSizes' );
-	if ( settingsSizes ) {
-		spacingSizes.push( ...settingsSizes );
-	}
+	const maybeDefaultSizes =
+		defaultSpacingSizesEnabled !== false ? defaultSizes : null;
 
-	if ( spacingSizes.length > 8 ) {
-		spacingSizes.unshift( {
-			name: __( 'Default' ),
-			slug: 'default',
-			size: undefined,
-		} );
-	}
+	return useMemo( () => {
+		const zeroSize = [ { name: 0, slug: '0', size: 0 } ];
 
-	return spacingSizes;
+		const sizesLength =
+			zeroSize.length +
+			( customSizes?.length ?? 0 ) +
+			( themeSizes?.length ?? 0 ) +
+			( maybeDefaultSizes?.length ?? 0 );
+
+		const maybeUnsetSize =
+			sizesLength > RANGE_CONTROL_MAX_SIZE
+				? [
+						{
+							name: __( 'Default' ),
+							slug: 'default',
+							size: undefined,
+						},
+				  ]
+				: null;
+
+		return [
+			...maybeUnsetSize,
```

Comment:
I'm getting `Uncaught TypeError: maybeUnsetSize is null` when I insert a spacer in the post editor. `sizesLength` for me is 7 and my theme.json uses:
```

"spacing": {
	"spacingScale": {
		"steps": 0
	},
	"spacingSizes": [
		{
			"name": "1",
			"size": "1rem",
			"slug": "10"
		},
		{
			"name": "2",
			"size": "2rem",
			"slug": "20"
		},
		{
			"name": "3",
			"size": "3rem",
			"slug": "30"
		},
		{
			"name": "4",
			"size": "4rem",
			"slug": "40"
		},
		{
			"name": "5",
			"size": "5rem",
			"slug": "50"
		},
		{
			"name": "6",
			"size": "6rem",
			"slug": "60"
		}
	],
	"units": ["%", "px", "em", "rem", "vh", "vw"]
},
```